In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
df = pd.read_csv('/home/joseph/Desktop/ml lab/lab1/dataset/Melbourne_housing_FULL.csv')
df.head(20)

Suburb              Address  Rooms Type      Price Method  SellerG  \
0   Abbotsford        68 Studley St      2    h        NaN     SS   Jellis   
1   Abbotsford         85 Turner St      2    h  1480000.0      S   Biggin   
2   Abbotsford      25 Bloomburg St      2    h  1035000.0      S   Biggin   
3   Abbotsford   18/659 Victoria St      3    u        NaN     VB   Rounds   
4   Abbotsford         5 Charles St      3    h  1465000.0     SP   Biggin   
5   Abbotsford     40 Federation La      3    h   850000.0     PI   Biggin   
6   Abbotsford          55a Park St      4    h  1600000.0     VB   Nelson   
7   Abbotsford         16 Maugie St      4    h        NaN     SN   Nelson   
8   Abbotsford         53 Turner St      2    h        NaN      S   Biggin   
9   Abbotsford         99 Turner St      2    h        NaN      S  Collins   
10  Abbotsford       129 Charles St      2    h   941000.0      S   Jellis   
11  Abbotsford         124 Yarra St      3    h  1876000.0      S   Nelson   
12  Abbotsford  121/56 Nicholson St      2    u        NaN     PI   Biggin   
13  Abbotsford        17 Raphael St      4    h        NaN      W   Biggin   
14  Abbotsford        98 Charles St      2    h  1636000.0      S   Nelson   
15  Abbotsford     217 Langridge St      3    h  1000000.0      S   Jellis   
16  Abbotsford      18a Mollison St      2    t   745000.0      S   Jellis   
17  Abbotsford   6/241 Nicholson St      1    u   300000.0      S   Biggin   
18  Abbotsford        10 Valiant St      2    h  1097000.0      S   Biggin   
19  Abbotsford  403/609 Victoria St      2    u   542000.0      S   Dingle   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0   3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1   3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2   4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3   4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4   4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
5   4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
6   4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
7   6/08/2016       2.5    3067.0  ...       2.0  2.0     400.0         220.0   
8   6/08/2016       2.5    3067.0  ...       1.0  2.0     201.0           NaN   
9   6/08/2016       2.5    3067.0  ...       2.0  1.0     202.0           NaN   
10  7/05/2016       2.5    3067.0  ...       1.0  0.0     181.0           NaN   
11  7/05/2016       2.5    3067.0  ...       2.0  0.0     245.0         210.0   
12  7/11/2016       2.5    3067.0  ...       2.0  1.0    4292.0          82.0   
13  7/11/2016       2.5    3067.0  ...       2.0  0.0     230.0         147.0   
14  8/10/2016       2.5    3067.0  ...       1.0  2.0     256.0         107.0   
15  8/10/2016       2.5    3067.0  ...       NaN  NaN       NaN           NaN   
16  8/10/2016       2.5    3067.0  ...       NaN  NaN       NaN           NaN   
17  8/10/2016       2.5    3067.0  ...       1.0  1.0       0.0           NaN   
18  8/10/2016       2.5    3067.0  ...       1.0  2.0     220.0          75.0   
19  8/10/2016       2.5    3067.0  ...       NaN  NaN       NaN           NaN   

    YearBuilt         CouncilArea Lattitude  Longtitude  \
0         NaN  Yarra City Council  -37.8014    144.9958   
1         NaN  Yarra City Council  -37.7996    144.9984   
2      1900.0  Yarra City Council  -37.8079    144.9934   
3         NaN  Yarra City Council  -37.8114    145.0116   
4      1900.0  Yarra City Council  -37.8093    144.9944   
5         NaN  Yarra City Council  -37.7969    144.9969   
6      2014.0  Yarra City Council  -37.8072    144.9941   
7      2006.0  Yarra City Council  -37.7965    144.9965   
8      1900.0  Yarra City Council  -37.7995    144.9974   
9      1900.0  Yarra City Council  -37.7996    144.9989   
10        NaN  Yar

In [4]:
uniqueCounts = df.nunique();
print("Unique count across columns:")
print(uniqueCounts);

Unique count across columns:
Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64


In [5]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

In [6]:
df['Price'].fillna(value = df.Price.mean(), inplace = True)
df['Distance'].fillna(value = df.Distance.mean(), inplace = True)
df['Postcode'].fillna(value = df.Postcode.mean(), inplace = True)
df['Bedroom2'].fillna(value = df.Bedroom2.mean(), inplace = True)
df['Bathroom'].fillna(value = df.Bathroom.mean(), inplace = True)
df['Car'].fillna(value = df.Car.mean(), inplace = True)
df['Landsize'].fillna(value = df.Landsize.mean(), inplace = True)
df['Bedroom2'].fillna(value = df.Bedroom2.mean(), inplace = True)
df['YearBuilt'].fillna(value = df.YearBuilt.mean(), inplace = True)
df['Lattitude'].fillna(value = df.Lattitude.mean(), inplace = True)
df['Longtitude'].fillna(value = df.Longtitude.mean(), inplace = True)
df['Propertycount'].fillna(value = df.Propertycount.mean(), inplace = True)
df['BuildingArea'].fillna(value = df.BuildingArea.mean(), inplace = True)


In [7]:
df.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
CouncilArea      3
Lattitude        0
Longtitude       0
Regionname       3
Propertycount    0
dtype: int64

In [8]:
df = df.fillna(0)

In [9]:
df

Suburb             Address  Rooms Type         Price Method  \
0      Abbotsford       68 Studley St      2    h  1.050173e+06     SS   
1      Abbotsford        85 Turner St      2    h  1.480000e+06      S   
2      Abbotsford     25 Bloomburg St      2    h  1.035000e+06      S   
3      Abbotsford  18/659 Victoria St      3    u  1.050173e+06     VB   
4      Abbotsford        5 Charles St      3    h  1.465000e+06     SP   
...           ...                 ...    ...  ...           ...    ...   
34852  Yarraville         13 Burns St      4    h  1.480000e+06     PI   
34853  Yarraville       29A Murray St      2    h  8.880000e+05     SP   
34854  Yarraville      147A Severn St      2    t  7.050000e+05      S   
34855  Yarraville    12/37 Stephen St      3    h  1.140000e+06     SP   
34856  Yarraville    3 Tarrengower St      2    h  1.020000e+06     PI   

             SellerG        Date  Distance  Postcode  ...  Bathroom       Car  \
0             Jellis   3/09/2016       2.5    3067.0  ...  1.000000  1.000000   
1             Biggin   3/12/2016       2.5    3067.0  ...  1.000000  1.000000   
2             Biggin   4/02/2016       2.5    3067.0  ...  1.000000  0.000000   
3             Rounds   4/02/2016       2.5    3067.0  ...  2.000000  1.000000   
4             Biggin   4/03/2017       2.5    3067.0  ...  2.000000  0.000000   
...              ...         ...       ...       ...  ...       ...       ...   
34852            Jas  24/02/2018       6.3    3013.0  ...  1.000000  3.000000   
34853        Sweeney  24/02/2018       6.3    3013.0  ...  2.000000  1.000000   
34854            Jas  24/02/2018       6.3    3013.0  ...  1.000000  2.000000   
34855  hockingstuart  24/02/2018       6.3    3013.0  ...  1.624798  1.728845   
34856             RW  24/02/2018       6.3    3013.0  ...  1.000000  0.000000   

         Landsize  BuildingArea    YearBuilt               CouncilArea  \
0      126.000000      160.2564  1965.289885        Yarra City Council   
1      202.000000      160.2564  1965.289885        Yarra City Council   
2      156.000000       79.0000  1900.000000        Yarra City Council   
3        0.000000      160.2564  1965.289885        Yarra City Council   
4      134.000000      150.0000  1900.000000        Yarra City Council   
...           ...           ...          ...                       ...   
34852  593.000000      160.2564  1965.289885  Maribyrnong City Council   
34853   98.000000      104.0000  2018.000000  Maribyrnong City Council   
34854  220.000000      120.0000  2000.000000  Maribyrnong City Council   
34855  593.598993      160.2564  1965.289885  Maribyrnong City Council   
34856  250.000000      103.0000  1930.000000  Maribyrnong City Council   

       Lattitude  Longtitude             Regionname Propertycount  
0     -37.801400  144.995800  Northern Metropolitan        4019.0  
1     -37.799600  144.998400  Northern Metropolitan        4019.0  
2     -37.807900  144.993400  Northern Metropolitan        4019.0  
3     -37.811400  145.011600  Northern Metropolitan        4019.0  
4     -37.809300  144.994400  Northern Metropolitan        4019.0  
...          ...         ...                    ...           ...  
34852 -37.810530  144.884670   Western Metropolitan        6543.0  
34853 -37.815510  144.888260   Western Metropolitan        6543.0  
34854 -37.822860  144.878560   Western Metropolitan        6543.0  
34855 -37.810634  145.001851   Western Metropolitan        6543.0  
34856 -37.818100  144.893510   Western Metropolitan        6543.0  

[34857 rows x 21 columns]

In [10]:
df.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  34857.000000  3.485700e+04  34857.000000  34857.000000  34857.000000   
mean       3.031012  1.050173e+06     11.184929   3116.062859      3.084647   
std        0.969933  5.671357e+05      6.788795    109.022339      0.857337   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.950000e+05      6.400000   3051.000000      3.000000   
50%        3.000000  1.050173e+06     10.300000   3103.000000      3.000000   
75%        4.000000  1.150000e+06     14.000000   3156.000000      3.084647   
max       16.000000  1.120000e+07     48.100000   3978.000000     30.000000   

           Bathroom           Car       Landsize  BuildingArea     YearBuilt  \
count  34857.000000  34857.000000   34857.000000  34857.000000  34857.000000   
mean       1.624798      1.728845     593.598993    160.256400   1965.289885   
std        0.633013      0.875119    2763.694121    251.943934     24.932350   
min        0.000000      0.000000       0.000000      0.000000   1196.000000   
25%        1.000000      1.000000     357.000000    160.000000   1965.289885   
50%        1.624798      1.728845     593.598993    160.256400   1965.289885   
75%        2.000000      2.000000     598.000000    160.256400   1965.289885   
max       12.000000     26.000000  433014.000000  44515.000000   2106.000000   

          Lattitude    Longtitude  Propertycount  
count  34857.000000  34857.000000   34857.000000  
mean     -37.810634    145.001851    7572.888306  
std        0.079280      0.105528    4427.899750  
min      -38.190430    144.423790      83.000000  
25%      -37.846900    144.964400    4385.000000  
50%      -37.810634    145.001851    6763.000000  
75%      -37.770900    145.051750   10412.000000  
max      -37.390200    145.526350   21650.000000

In [11]:
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']


In [14]:
x=df[melbourne_features]


In [16]:
x.describe()


Rooms      Bathroom       Landsize     Lattitude    Longtitude
count  34857.000000  34857.000000   34857.000000  34857.000000  34857.000000
mean       3.031012      1.624798     593.598993    -37.810634    145.001851
std        0.969933      0.633013    2763.694121      0.079280      0.105528
min        1.000000      0.000000       0.000000    -38.190430    144.423790
25%        2.000000      1.000000     357.000000    -37.846900    144.964400
50%        3.000000      1.624798     593.598993    -37.810634    145.001851
75%        4.000000      2.000000     598.000000    -37.770900    145.051750
max       16.000000     12.000000  433014.000000    -37.390200    145.526350

In [17]:
y=df.Price

In [18]:
y.describe()


count    3.485700e+04
mean     1.050173e+06
std      5.671357e+05
min      8.500000e+04
25%      6.950000e+05
50%      1.050173e+06
75%      1.150000e+06
max      1.120000e+07
Name: Price, dtype: float64

In [27]:
from sklearn.tree import DecisionTreeRegressor

# Defining model
melbourne_model = DecisionTreeRegressor(random_state=42)

# Fit model
melbourne_model.fit(x, y)

DecisionTreeRegressor(random_state=42)

In [22]:
print("Making predictions for the following 5 houses:")
print(x.head())
print("The predictions are")
print(melbourne_model.predict(x.head()))

Making predictions for the following 5 houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
0      2       1.0     126.0   -37.8014    144.9958
1      2       1.0     202.0   -37.7996    144.9984
2      2       1.0     156.0   -37.8079    144.9934
3      3       2.0       0.0   -37.8114    145.0116
4      3       2.0     134.0   -37.8093    144.9944
The predictions are
[1050173.34495541 1480000.         1035000.         1050173.34495541
 1465000.        ]


In [24]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melbourne_model.predict(x)
mean_absolute_error(y, predicted_home_prices)

70193.33202066459

**This means that the model is off by $70193 on house prices**

In [25]:
from sklearn.model_selection import train_test_split


train_X, val_X, train_y, val_y = train_test_split(x, y, random_state = 42)
# Define model
melbourne_model = DecisionTreeRegressor()
# Fit model
melbourne_model.fit(train_X, train_y)

# get predicted prices on validation data
val_predictions = melbourne_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

317819.81205841387
